From 670b8f4ed99d40a2a4d76647f9b3f2bf46c731fd Mon Sep 17 00:00:00 2001
From: Yelena <yelena@marvell.com>
Date: Mon, 15 Aug 2016 09:05:39 +0300
Subject: [PATCH 0427/1345] pp3: don't use SW queues 8-15 in frame 2 for data
 path

commit  5148151d26a768b486ee3f2d539fa5345832758c from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

There are only 40 RU classes allocated for HMAC
So for data path only 40 TX SWQs can be used
HFrame #0: TXQs 0..15
HFrame #1: TXQs 16..31
HFrame #2: TXQs 32..39

Change-Id: Idf53bdd3e0a26bed2de1ddf1d4940fbcd0845b7f
Signed-off-by: Yelena <yelena@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/31933
Reviewed-by: Yan Markman <ymarkman@marvell.com>
Reviewed-by: Dmitri Epshtein <dima@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../ethernet/marvell/pp3/platform/mv_pp3_config.c  |   13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/marvell/pp3/platform/mv_pp3_config.c b/drivers/net/ethernet/marvell/pp3/platform/mv_pp3_config.c
index 5bb6cfe..aa2042e 100644
--- a/drivers/net/ethernet/marvell/pp3/platform/mv_pp3_config.c
+++ b/drivers/net/ethernet/marvell/pp3/platform/mv_pp3_config.c
@@ -38,6 +38,7 @@ struct mv_pp3_net_if_cfg {
 };
 
 struct mv_pp3_cpu_cfg {
+	u32 sw_max_q_num[MV_PP3_HFRM_NUM];
 	u32 sw_free_rxq[MV_PP3_HFRM_NUM];	/* free HMAC RX SW queue number per frame */
 	u32 sw_free_txq[MV_PP3_HFRM_NUM];	/* free HMAC TX SW queue number per frame */
 	u32 free_irq_group[MV_PP3_HFRM_NUM];	/* free HMAC IRQ group number per frame */
@@ -87,8 +88,13 @@ void mv_pp3_configurator_init(struct mv_pp3 *priv)
 	/* first internal BP group */
 	mv_pp3_cfg_sw_info.bp_group_id = 1;
 
-	for (i = 0; i < MV_PP3_HFRM_NUM; i++)
+	for (i = 0; i < MV_PP3_HFRM_NUM; i++) {
 		mv_pp3_cfg_sw_info.frame_vip_vport[i] = -1;
+		if (i == 2)
+			mv_pp3_cfg_sw_info.sw_max_q_num[i] = 8;
+		else
+			mv_pp3_cfg_sw_info.sw_max_q_num[i] = MV_PP3_HFRM_Q_NUM;
+	}
 
 	/* init configurator clients */
 	for (i = 0; i < PP3_CLIENTS_NUM; i++) {
@@ -795,8 +801,9 @@ int mv_pp3_cfg_dp_reserve_txq(int id, int if_num, int cpu, int q_num)
 		/* sort all frames per free RXQs number */
 		mv_pp3_cfg_sort(mv_pp3_cfg_sw_info.sw_free_txq, ind_arr);
 
-		for (i = 0; (i < MV_PP3_HFRM_NUM) && !found; i++) {
-			if ((mv_pp3_cfg_sw_info.sw_free_txq[ind_arr[i]] + q_num) > MV_PP3_HFRM_Q_NUM)
+		/* Don't use HF3 and HF2-Q8..15 for Data Path TXQs - No enough RU classes */
+		for (i = 0; (i < (MV_PP3_HFRM_NUM - 1)) && !found; i++) {
+			if ((mv_pp3_cfg_sw_info.sw_free_txq[ind_arr[i]] + q_num) >= mv_pp3_cfg_sw_info.sw_max_q_num[i])
 				continue;
 
 			/* each emac interface placed in different frame */
-- 
1.7.9.5

